File: web/src/lib/api/inspections.ts
import { getSupabase } from "@/lib/supabase";
import { apiFetch } from "./client";
import { InspectionListApiResponse, InspectionListParams } from "@/types/api";
import { InspectionDetail, InspectionListItem } from "@/types/models";

export async function getInspections(params: InspectionListParams = {}): Promise<InspectionListApiResponse> {
  try {
    const supabase = getSupabase();
    let query = supabase.from("inspections").select("*, defects(*)", { count: "exact" });

    if (params.status) {
      query = query.eq("status", params.status);
    }
    if (params.model) {
      query = query.eq("model", params.model);
    }
    if (params.station_id) {
      query = query.eq("station_id", params.station_id);
    }
    if (params.review_status) {
      query = query.eq("review_status", params.review_status);
    }
    if (params.date_from) {
      query = query.gte("captured_at", params.date_from);
    }
    if (params.date_to) {
      query = query.lte("captured_at", params.date_to);
    }
    if (params.search) {
      query = query.or(`source.ilike.%${params.search}%,station_id.ilike.%${params.search}%,model.ilike.%${params.search}%`);
    }

    const limit = params.limit ?? 20;
    const offset = params.offset ?? 0;
    query = query.order("captured_at", { ascending: false }).range(offset, offset + limit - 1);

    const { data, count, error } = await query;
    if (error) throw error;

    const items: InspectionListItem[] = (data || []).map((row: any) => ({
      id: row.id,
      captured_at: row.captured_at,
      model: row.model || "yolov8s",
      status: row.status,
      defect_count: Array.isArray(row.defects) ? row.defects.length : (row.defect_count ?? 0),
      image_url: row.image_url || row.image_path || "",
      station_id: row.station_id || "STATION-01",
      review_status: row.review_status || "UNREVIEWED",
      final_status: row.final_status || row.status,
    }));

    let filteredItems = items;
    if (params.defect_class) {
      filteredItems = items.filter((_, idx) => {
        const defects = data?.[idx]?.defects || [];
        return defects.some((d: any) => (d.class || d.cls) === params.defect_class);
      });
    }

    return {
      items: filteredItems,
      total: count ?? filteredItems.length,
    };
  } catch (err) {
    const searchParams = new URLSearchParams();
    if (params.status) searchParams.set("status", params.status);
    if (params.model) searchParams.set("model", params.model);
    if (params.search) searchParams.set("search", params.search);
    if (params.limit !== undefined) searchParams.set("limit", params.limit.toString());
    if (params.offset !== undefined) searchParams.set("offset", params.offset.toString());
    const q = searchParams.toString();
    return apiFetch<InspectionListApiResponse>(`/inspections${q ? `?${q}` : ""}`);
  }
}

export async function getInspectionDetail(id: string): Promise<InspectionDetail> {
  try {
    const supabase = getSupabase();
    const { data, error } = await supabase
      .from("inspections")
      .select("*, defects(*), reviews(*)")
      .eq("id", id)
      .single();

    if (error || !data) throw error || new Error("Inspection not found");

    const defects = (data.defects || []).map((d: any) => {
      const box = Array.isArray(d.box_2d) ? d.box_2d : [0, 0, 0, 0];
      return {
        id: d.id,
        class: d.class || d.cls || "defect",
        confidence: d.confidence ?? 0.9,
        severity: d.severity || "Moderate",
        box_x1: d.box_x1 ?? box[0] ?? 0,
        box_y1: d.box_y1 ?? box[1] ?? 0,
        box_x2: d.box_x2 ?? box[2] ?? 0,
        box_y2: d.box_y2 ?? box[3] ?? 0,
      };
    });

    const reviews = (data.reviews || []).map((r: any) => ({
      id: r.id,
      inspection_id: r.inspection_id,
      automated_result: data.status,
      review_decision: r.review_decision || r.status || "CONFIRM",
      final_result: r.final_result || data.status,
      reviewer_email: r.reviewer_email || r.inspector_email || "reviewer@example.com",
      justification: r.justification || r.notes || "",
      notes: r.notes,
      created_at: r.reviewed_at || r.created_at || new Date().toISOString(),
    }));

    return {
      id: data.id,
      captured_at: data.captured_at,
      source: data.source || data.source_image_name || "capture.jpg",
      image_url: data.image_url || data.image_path || "",
      annotated_url: data.annotated_url || data.annotated_image_path || data.image_url || "",
      model: data.model || "yolov8s",
      status: data.status,
      station_id: data.station_id || "STATION-01",
      created_at: data.captured_at,
      defects,
      reviews,
      review_status: data.review_status || "UNREVIEWED",
      final_status: data.final_status || data.status,
    };
  } catch {
    return apiFetch<InspectionDetail>(`/inspections/${id}`);
  }
}

File: PCB-Fault-Detection-System-main/backend/app/routers/inspections.py
import json
import uuid
from datetime import datetime

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile, status
from sqlalchemy import func, or_, select
from sqlalchemy.orm import Session, selectinload

from ..auth import get_current_user
from ..db import get_db
from ..models import AuditLog, Defect, Inspection, InspectionStatus, Notification, Severity, User
from ..schemas import InspectionDetailResponse, InspectionListItem, InspectionListResponse
from ..storage import upload_file

router = APIRouter(prefix="/inspections", tags=["inspections"])


def _parse_result(value: str) -> dict:
    try:
        data = json.loads(value)
        if not isinstance(data, dict) or not isinstance(data.get("defects", []), list):
            raise ValueError
        if data.get("status") not in {"PASS", "FAIL"} or not isinstance(data.get("model"), str):
            raise ValueError
        return data
    except (json.JSONDecodeError, ValueError, TypeError):
        raise HTTPException(status_code=422, detail="result must be valid inspection JSON")


@router.post("/ingest", status_code=status.HTTP_201_CREATED)
def ingest(
    captured_at: str = Form(...),
    source: str = Form(...),
    result: str = Form(...),
    image: UploadFile = File(...),
    annotated: UploadFile = File(...),
    station_id: str | None = Form(None),
    local_id: int | None = Form(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    payload = _parse_result(result)
    try:
        captured = datetime.fromisoformat(captured_at.replace("Z", "+00:00"))
    except ValueError:
        raise HTTPException(status_code=422, detail="captured_at must be ISO 8601")

    # --- Idempotency Check ---
    # Check if this exact inspection has already been uploaded from this station
    if local_id is not None and station_id:
        existing = db.scalar(
            select(Inspection).where(
                Inspection.local_id == local_id,
                Inspection.station_id == station_id,
            )
        )
        if existing:
            return {"id": str(existing.id), "status": "exists"}

    # Secondary check: same source and exact captured timestamp
    existing_by_source = db.scalar(
        select(Inspection).where(
            Inspection.source == source,
            Inspection.captured_at == captured,
        )
    )
    if existing_by_source:
        return {"id": str(existing_by_source.id), "status": "exists"}

    inspection_id = uuid.uuid4()
    try:
        image_url = upload_file(image, f"inspections/{inspection_id}.jpg")
        annotated_url = upload_file(annotated, f"inspections/{inspection_id}_annotated.jpg")

        has_critical = False
        defects_to_add = []
        for defect in payload["defects"]:
            box = defect.get("box")
            if not isinstance(box, list) or len(box) != 4:
                raise ValueError("Each defect needs a four-value box")
            sev = Severity(defect["severity"])
            if sev == Severity.Critical:
                has_critical = True
            defects_to_add.append(
                Defect(
                    inspection_id=inspection_id,
                    class_name=str(defect["class"]),
                    confidence=float(defect["confidence"]),
                    severity=sev,
                    box_x1=float(box[0]),
                    box_y1=float(box[1]),
                    box_x2=float(box[2]),
                    box_y2=float(box[3]),
                )
            )

        review_status = "PENDING" if (has_critical or payload["status"] == "FAIL") else "UNREVIEWED"

        inspection = Inspection(
            id=inspection_id,
            captured_at=captured,
            source=source,
            model=payload["model"],
            status=InspectionStatus(payload["status"]),
            image_url=image_url,
            annotated_url=annotated_url,
            station_id=station_id or "STATION-01",
            local_id=local_id,
            review_status=review_status,
        )
        db.add(inspection)
        for d in defects_to_add:
            db.add(d)

        # Automated notification on critical defect
        if has_critical:
            db.add(
                Notification(
                    category="Critical Defect",
                    title=f"Critical Defect on {source}",
                    message=f"Station {station_id or 'STATION-01'} detected critical defect(s) using model {payload['model']}.",
                    severity="Critical",
                    inspection_id=inspection_id,
                )
            )

        # Audit log for inspection creation
        db.add(
            AuditLog(
                user_id=current_user.id,
                user_email=current_user.email,
                action="INSPECTION_INGESTED",
                entity="inspection",
                entity_id=str(inspection_id),
                description=f"Ingested inspection {inspection_id} from {station_id or 'STATION-01'} ({source}) - Result: {payload['status']}",
                details_json={"model": payload["model"], "defects": len(defects_to_add), "has_critical": has_critical},
            )
        )

        db.commit()
    except (KeyError, TypeError, ValueError) as exc:
        db.rollback()
        raise HTTPException(status_code=422, detail=f"Invalid defect data: {exc}")
    except Exception as exc:
        db.rollback()
        raise HTTPException(status_code=500, detail=f"Could not upload and store inspection: {exc}")

    return {"id": str(inspection_id), "status": "created"}


@router.get("", response_model=InspectionListResponse)
def list_inspections(
    status: InspectionStatus | None = None,
    model: str | None = None,
    date_from: datetime | None = None,
    date_to: datetime | None = None,
    defect_class: str | None = None,
    station_id: str | None = None,
    review_status: str | None = None,
    search: str | None = None,
    limit: int = 50,
    offset: int = 0,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    if not 1 <= limit <= 200 or offset < 0:
        raise HTTPException(status_code=422, detail="limit must be 1–200 and offset must not be negative")

    query = select(Inspection)
    if status:
        query = query.where(Inspection.status == status)
    if model:
        query = query.where(Inspection.model == model)
    if station_id:
        query = query.where(Inspection.station_id == station_id)
    if review_status:
        query = query.where(Inspection.review_status == review_status)
    if date_from:
        query = query.where(Inspection.captured_at >= date_from)
    if date_to:
        query = query.where(Inspection.captured_at <= date_to)
    if defect_class:
        query = query.where(Inspection.defects.any(Defect.class_name == defect_class))
    if search:
        search_term = f"%{search.strip()}%"
        query = query.where(
            or_(
                Inspection.source.ilike(search_term),
                Inspection.model.ilike(search_term),
                Inspection.station_id.ilike(search_term),
            )
        )

    total = db.scalar(select(func.count()).select_from(query.subquery())) or 0
    inspections = db.scalars(
        query.options(selectinload(Inspection.defects))
        .order_by(Inspection.captured_at.desc(), Inspection.id.desc())
        .offset(offset)
        .limit(limit)
    ).all()

    return InspectionListResponse(
        total=total,
        items=[
            InspectionListItem(
                id=i.id,
                captured_at=i.captured_at,
                status=i.status,
                model=i.model,
                defect_count=len(i.defects),
                image_url=i.image_url,
                station_id=i.station_id,
                review_status=i.review_status,
                final_status=i.final_status,
            )
            for i in inspections
        ],
    )


@router.get("/{inspection_id}", response_model=InspectionDetailResponse)
def inspection_detail(
    inspection_id: uuid.UUID,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    inspection = db.scalar(
        select(Inspection)
        .options(selectinload(Inspection.defects), selectinload(Inspection.reviews))
        .where(Inspection.id == inspection_id)
    )
    if inspection is None:
        raise HTTPException(status_code=404, detail="Inspection not found")
    return inspection
